de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java

```diff
@@ -135,6 +139,12 @@ public void negotiationTerminated(final SessionNegotiation negotiation) {
         @Override
         public void negotiationTerminated(final ProjectNegotiation negotiation) {
             currentProjectNegotiations.remove(negotiation);
+
+            if (currentProjectNegotiations.isEmpty()) {
+                synchronized (nextProjectNegotiation) {
```

Comment:
Why isn't this behavior moved to the class instead, e.g creating a signal method. The code is getting too complicated. If I count correctly there a now 3 locks involved for the logic. I think we should prefer an asynchronous solution instead of a concurrent one, e.g:

collector.add(resources)
excecutor.submit(runSharing)

the runSharing runnable can then look if there is a running negotiation, if yes abort. if there are no resources, abort

This is somewhat easier to handle, e.g you can execute the code 1 time or 1000 times it will not matter, the result be the same.